edgedb/errors/_base.py

```diff
@@ -180,15 +226,83 @@ def _severity_name(severity):
     return 'PANIC'
 
 
+def _format_error(msg, query, start, offset, line, col, hint):
+    c = get_color()
+    rv = io.StringIO()
+    rv.write(f"{c.BOLD}{msg}{c.ENDC}{LINESEP}")
+    lines = query.splitlines(keepends=True)
+    num_len = len(str(len(lines)))
+    rv.write(f"{c.BLUE}{'':>{num_len}} ┌─{c.ENDC} query:{line}:{col}{LINESEP}")
+    rv.write(f"{c.BLUE}{'':>{num_len}} │ {c.ENDC}{LINESEP}")
+    for num, line in enumerate(lines):
+        length = len(line)
+        line = line.rstrip()  # we'll use our own line separator
+        if start >= length:
+            # skip lines before the error
+            start -= length
+            continue
+
+        if start >= 0:
+            # Error starts in current line, write the line before the error
+            first_half = repr(line[:start])[1:-1]
+            line = line[start:]
+            length -= start
+            rv.write(f"{c.BLUE}{num + 1:>{num_len}} │   {c.ENDC}{first_half}")
+            start = _unicode_width(first_half)
+        else:
+            # Multi-line error continues
+            rv.write(f"{c.BLUE}{num + 1:>{num_len}} │ {c.FAIL}│ {c.ENDC}")
+
+        if offset > length:
+            # Error is ending beyond current line
+            line = repr(line)[1:-1]
+            rv.write(f"{c.FAIL}{line}{c.ENDC}{LINESEP}")
+            if start >= 0:
+                # Multi-line error starts
+                rv.write(f"{c.BLUE}{'':>{num_len}} │ "
+                         f"{c.FAIL}╭─{'─' * start}^{c.ENDC}{LINESEP}")
+            offset -= length
+            start = -1  # mark multi-line
+        else:
+            # Error is ending within current line
+            first_half = repr(line[:offset])[1:-1]
+            line = repr(line[offset:])[1:-1]
+            rv.write(f"{c.FAIL}{first_half}{c.ENDC}{line}{LINESEP}")
+            size = _unicode_width(first_half)
+            if start >= 0:
+                # Mark single-line error
+                rv.write(f"{c.BLUE}{'':>{num_len}} │   {' ' * start}"
+                         f"{c.FAIL}{'^' * size} {hint}{c.ENDC}")
+            else:
+                # End of multi-line error
+                rv.write(f"{c.BLUE}{'':>{num_len}} │ "
+                         f"{c.FAIL}╰─{'─' * (size - 1)}^ {hint}{c.ENDC}")
+            break
+    return rv.getvalue()
+
+
+def _unicode_width(text):
```

Comment:
Note that this is an approximation that will overestimate the length of strings using combining characters that *don't* normalize away.
This means we can get stuff like
```
cannot redefine property 'x' of object type 'std::FreeObject' as scalar type 'std::str'
  ┌─ query:2:26
  │ 
2 │   select { x := 1 } { x := 'f̷͈͎͒̕ǫ̴̏͌ö̶̱̘' };
  │                            ^^^^^^^^^^^^^^^^ error
```

https://zalgo.org/ has a great generator for pathological unicode strings.

Probably this is almost always fine.

Reply:
The CLI's error formatter does seem to handle this one right. I'm not sure exactly what algorithm it uses.
I think to some extent anything we do is an approximation because it depends on assumptions about how the terminal will render it.

Unicode does have a notion of "grapheme clusters", and there are python libraries for handling them (https://github.com/alvinlindstam/grapheme). I'm not sure that this is the best approach either, though, since the width of those clusters can vary and I don't know if there is good way to guess that well. (Consider `🇨🇦`, which depending on your terminal will show as either a flag or the letters CA, but either way will probably take two characters, but is only one grapheme cluster.)

Maybe the best approach on our side would be basically just to drop combining characters.

There is a potentially infinitely deep rabbit hole here.

Reply:
Rust uses de facto standard in Rust: https://crates.io/crates/unicode-width
Which uses Unicode Standard Annex 11 for calculating width.

Note: unicode width != number of graphemes. As width of a grapheme can be two chars.

And you're right that in generic case it's impossible to calculate width. Because some characters can have width not just depending on the terminal but also on the specific font (if there is a character for a specific set of emojis). `unicode-width`'s frontpage gives an example.

# Dropping Zero-width Joiners

1. I've tested three terminals I have around (xfce4, alacritty, wezterm), and all of them do match width with unicode width, even though some of them do not render some of the characters (i.e. flags are only supported in wezterm). *Disclaimer: alacritty and wezterm are implemented in Rust so this explains it a bit, although xfce4 is not and has much longer history.*
2. Copying from any of the three terminal's I've tried does discard zero-width joiner by itself
3. Rendering in Firefox on my machine gives invalid width regardless of if there is a zero-width joiner (i.e. it somehow draws emojis 1.5 of character width :shrug: )
4. Rendering in Chrome on my machine gives good results when zero-width joiner is dropped (like normally copying from terminal), but doesn't if I somehow manage to keep that character (i.e. piping output to `xsel -b` manually).

Note: zalgo-generated stuff looks like plain `foo` in all three terminals (although those unicode tricks are present in the output)

The summary of this: dropping zero-width joiners is safe to do, although looks unnecessary in my setup. If that's not true for some popular terminals or on other systems it could be done.

# Fixup

For my examples this function works:
```
def unicode_width(s):
    return sum(0 if unicodedata.category(c) in ('Mn', 'Cf') else
               2 if unicodedata.east_asian_width(c) == "W" else 1
        for c in s)
```
But I'm not sure if this is good enough. The libraries doing this (rustic unicode-width, and pythonic urwid) have their own width tables.

Reply:
Added in #404